src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/SemanticTokensCache.cs

```diff
@@ -0,0 +1,210 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Razor;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    /// <summary>
+    /// Caches tokens on a per-line basis for documents based on semantic version.
+    /// </summary>
+    /// <remarks>
+    /// The cache makes the assumption that the tokens passed in are complete for the given range.
+    /// </remarks>
+    internal sealed class SemanticTokensCache
+    {
+        private const int MaxSemanticVersionPerDoc = 4;
+
+        // Nested cache mapping (URI -> (semanticVersion -> (line #s -> tokens on line)))
+        private readonly MemoryCache<DocumentUri, MemoryCache<VersionStamp, Dictionary<int, List<int>>>> _cache = new();
+
+        /// <summary>
+        /// Caches tokens on a per-line basis. If the given line has already been cached for the document
+        /// and its semantic version, the line's tokens will be ignored.
+        /// </summary>
+        /// <param name="uri">The URI associated with the passed-in tokens.</param>
+        /// <param name="semanticVersion">The document's semantic version associated with the passed-in tokens.</param>
+        /// <param name="range">The range associated with the passed-in tokens.</param>
+        /// <param name="tokens">The complete set of tokens for the passed-in range.</param>
+        internal void CacheTokens(
+            DocumentUri uri,
+            VersionStamp semanticVersion,
+            Range range,
+            int[] tokens)
+        {
+            if (!_cache.TryGetValue(uri, out var documentCache))
+            {
+                documentCache = new MemoryCache<VersionStamp, Dictionary<int, List<int>>>(sizeLimit: MaxSemanticVersionPerDoc);
+                _cache.Set(uri, documentCache);
+            }
+
+            if (!documentCache.TryGetValue(semanticVersion, out var lineToTokensDict))
+            {
+                lineToTokensDict = new Dictionary<int, List<int>>();
+                documentCache.Set(semanticVersion, lineToTokensDict);
+            }
+
+            var absoluteLine = 0;
```

Comment:
Actually, this makes me wonder if this will work.

Suppose we make two range requests for the same document, one for line 15, one for line 16. The first token for the first request will say that it's 15 lines from the start of the document. The request for line 16 will say that it's 16 lines from the start of the document, but will be listed immediately after the line 15 request, so instead that will be read as 16 lines _from line 15_, resulting in a location which is possibly outside the document, and definitely not correct.

The easiest solution to this I can think of would be to translate our results into absolute numbers before storage, and back to relative numbers on retrieval, but that's getting pretty complicated, and possibly expensive depending on the weight of the APIs we'd need to call.

Reply:
Yep, good point! This case is handled in the retrieval method. We could translate to absolute numbers before storage, but that seems like more work since we'd need to adjust the offsets again on retrieval anyway.